libs/langchain/src/agents/middlewareAgent/middleware/bigTool.ts

```diff
@@ -0,0 +1,239 @@
+ * @param options.maxRetries - Maximum number of retries if the model selects incorrect tools.
+ *
+ * @example
+ * ```ts
+ * import { llmToolSelectorMiddleware } from "langchain/agents/middleware";
+ *
+ * const middleware = llmToolSelectorMiddleware({
+ *   maxTools: 3,
+ *   systemPrompt: "Select the most relevant tools for the user's query."
+ * });
+ *
+ * const agent = createAgent({
+ *   model: "openai:gpt-4o",
+ *   tools: [tool1, tool2, tool3, tool4, tool5],
+ *   middleware: [middleware],
+ * });
+ * ```
+ */
+export function llmToolSelectorMiddleware(options: LLMToolSelectorConfig) {
+  return createMiddleware({
+    name: "LLMToolSelector",
+    contextSchema: LLMToolSelectorOptionsSchema,
+    async modifyModelRequest(request, _, runtime): Promise<ModelRequest> {
+      const model = runtime.context.model ?? options.model;
+      const maxTools = runtime.context.maxTools ?? options.maxTools;
+      const includeFullHistory =
+        runtime.context.includeFullHistory === DEFAULT_INCLUDE_FULL_HISTORY
+          ? options.includeFullHistory ?? runtime.context.includeFullHistory
+          : runtime.context.includeFullHistory ?? options.includeFullHistory;
+      const maxRetries =
+        runtime.context.maxRetries === DEFAULT_MAX_RETRIES
+          ? options.maxRetries ?? runtime.context.maxRetries
+          : runtime.context.maxRetries ?? options.maxRetries;
+      const defaultSystemPrompt =
+        runtime.context.systemPrompt === DEFAULT_SYSTEM_PROMPT
+          ? options.systemPrompt ?? runtime.context.systemPrompt
+          : runtime.context.systemPrompt ?? options.systemPrompt;
+
+      /**
+       * If no tools available, return request unchanged
+       */
+      if (!request.tools || request.tools.length === 0) {
+        return request;
+      }
+
+      /**
+       * Extract tool information
+       */
+      const toolInfo = runtime.tools.map((tool) => ({
+        name: tool.name as string,
+        description: tool.description,
+        tool,
+      }));
+
+      /**
+       * Build tool representation for the prompt
+       */
+      const toolRepresentation = toolInfo
+        .map(({ name, description }) => `- ${name}: ${description}`)
+        .join("\n");
+
+      /**
+       * Build system message
+       */
+      let systemMessage = `You are an agent that can use the following tools:\n${toolRepresentation}\n${defaultSystemPrompt}`;
+
+      if (includeFullHistory) {
+        const userMessages = request.messages
+          .filter(HumanMessage.isInstance)
+          .map((msg: BaseMessage) => msg.content)
+          .join("\n");
+        systemMessage += `\nThe full conversation history is:\n${userMessages}`;
+      }
+
+      if (maxTools !== undefined) {
+        systemMessage += ` You can select up to ${maxTools} tools.`;
+      }
+
+      /**
+       * Get the latest user message
+       */
+      const latestMessage = request.messages.at(-1);
```

Comment:
I believe that there's a bug here -- the latest message may not be a user message (it may be a ToolMessage).